### PR TITLE
Docker compose v2

### PR DESCRIPTION
**Pull request by 18601673727**
**Purpose:**
Solving https://github.com/sameersbn/docker-gitlab/pull/631 and https://github.com/sameersbn/docker-gitlab/issues/630.

**Known issue:**
- <del>SSH server seems not working, only be able to pull/push via HTTPS. </del>
- <del>HTTP(S) clone url which shown in the "project detail page" is not correct. (only with https://github.com/sameersbn/docker-gitlab/pull/658/commits/222267976b888ddd403161561f5f645761177ba6)</del>

### CODE REVIEW

**Comment by 18601673727** [comment]
Known issue:

SSH server seems not working, hasn't tested without proxy server yet.

File: docker-compose.yml [hunk 1]
```diff
@@ -8,12 +8,23 @@ postgresql:
...
+  image: sameersbn/redis:latest
+  volumes:
+    - /srv/docker/gitlab/redis:/var/lib/redis
+  networks:
+    - gitlab-tier
+
 gitlab:
   restart: always
   image: sameersbn/gitlab:8.6.4
   links:
```
Comment (SolidNerd) [inline docker-compose.yml]: Do we need the links anymore ? I guess it's redudant through the networks.

Comment (sameersbn) [inline docker-compose.yml]: yes. links are so yesterday. we need to get rid of them. Docker is in the process of retiring links and we should too :smile:

**Comment by 18601673727** [comment]
The workaround I've found is not stable enough, it only work 1 time after I `docker-compose up`, when stopped and re-run it, SSH server will not connectable again, sign.
```
git:(master) ssh -vvT gitlab@git.domain.com
OpenSSH_6.9p1, LibreSSL 2.1.8
debug1: Reading configuration data /Users/xxx/.ssh/config
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 21: Applying options for *
debug2: ssh_connect: needpriv 0
debug1: Connecting to git.domain.com [xxx.xxx.xxx.xxx] port 22.
debug1: connect to address xxx.xxx.xxx.xxx port 22: Operation timed out
ssh: connect to host git.domain.com port 22: Operation timed out
```

**Comment by sameersbn** [comment]
@18601673727 @SolidNerd have updated this PR in #661. Please review and test if possible.

**Comment by SolidNerd** [comment]
I will test it later

**Comment by 18601673727** [comment]
@sameersbn I'm afraid that SSH issue still at the same situation.

**Comment by sameersbn** [comment]
@18601673727 what exactly is the SSH issue you see? can you please share some details. I have tested SSH and HTTP urls/cloning without any issue.

**Comment by 18601673727** [comment]
@sameersbn  Well, gitlab runs behind a nginx proxy with ssl, SSH ports to 22, ubuntu14.04, everything fine when using compose V1.But your and mine compose V2 will both hang the command `git push -u origin master` or `ssh -vvT git@git.domain.com ` this is the issue I'm facing right now.

**Comment by sameersbn** [comment]
can you share a sample of the docker-compose.yml file you are using? preferable based on the one submitted in PR #661, as I have tested that one and found it to work.

Can you share the HTTPS/SSH clone urls which are not right and what you expect them to be.

Also some details of your docker setup? eg. the host os, whether docker-machine or vms are used for docker, docker version, etc.

P.S. the docker-compose.yml implemented in this PR is not really complaint with the version 2 spec. It has some very vital bits missing.

**Comment by 18601673727** [comment]
- As I said in my last comment, both file will cause SSH hanging. 
- URL not correct with 2222679 only 
- Docker and Compse version are newest.

Other information I'll offer you about 4 hours later, currently I'm ouside no computers to use.

Thanks!

**Comment by 18601673727** [comment]
Hi @sameersbn, here comes my docker-compose.yml:
```yml
version: '2'

services:

  postgresql:
    restart: always
    image: sameersbn/postgresql:9.4-18
    environment:
      - DB_USER=gitlab
      - DB_PASS=password
      - DB_NAME=gitlabhq_production
      - DB_EXTENSION=pg_trgm
    volumes:
      - /srv/docker/gitlab/postgresql:/var/lib/postgresql:Z
    networks:
      - gitlab-tier

  redis:
    restart: always
    image: sameersbn/redis:latest
    command:
    - --loglevel warning
    volumes:
      - /srv/docker/gitlab/redis:/var/lib/redis:Z
    networks:
      - gitlab-tier

  gitlab:
    restart: always
    image: sameersbn/gitlab:8.6.4
    ports:
      - "10080:80"
      - "22:22"
    depends_on:
      - redis
      - postgresql
    environment:
      - DB_HOST=postgresql
      - DB_USER=gitlab
      - DB_NAME=gitlabhq_production
      - DB_PASS=password
      - DB_ADAPTER=postgresql
      - REDIS_HOST=redis

      #- VIRTUAL_NETWORK=nginx-proxy
      - VIRTUAL_HOST=git.domain.com
      - LETSENCRYPT_HOST=git.domain.com
      - LETSENCRYPT_EMAIL=me@domain.com

      - DEBUG=false
      - TZ=Asia/Shanghai
      - GITLAB_TIMEZONE=Beijing

      - GITLAB_SECRETS_DB_KEY_BASE=long-and-random-alphanumeric-string

      - GITLAB_HOST=git.domain.com
      #- GITLAB_SSH_HOST=git.domain.com
      - GITLAB_PORT=443
      - GITLAB_SSH_PORT=22
      - GITLAB_HTTPS=true
      - GITLAB_RELATIVE_URL_ROOT=

      - GITLAB_NOTIFY_ON_BROKEN_BUILDS=true
      - GITLAB_NOTIFY_PUSHER=false

      - GITLAB_EMAIL=notifications@domain.com
      - GITLAB_EMAIL_REPLY_TO=noreply@domain.com
      - GITLAB_INCOMING_EMAIL_ADDRESS=reply@domain.com

      - GITLAB_BACKUP_SCHEDULE=daily
      - GITLAB_BACKUP_TIME=01:00

      - SMTP_ENABLED=false
      - SMTP_DOMAIN=mail.com
      - SMTP_HOST=smtp.mail.com
      - SMTP_PORT=465
      - SMTP_USER=git@domain.com
      - SMTP_PASS=xxxxxxxx
      - SMTP_STARTTLS=true
      - SMTP_AUTHENTICATION=login

      - IMAP_ENABLED=false
      - IMAP_HOST=imap.mail.com
      - IMAP_PORT=993
      - IMAP_USER=git@domain.com
      - IMAP_PASS=xxxxxxxx
      - IMAP_SSL=true
      - IMAP_STARTTLS=false
    volumes:
      - /srv/docker/gitlab/gitlab:/home/git/data:Z
    networks:
      - proxy-tier
      - gitlab-tier

networks:
  proxy-tier:
    external:
      name: nginx-proxy
  gitlab-tier:
    driver: bridge
```

```
$ docker version
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64
And some docker infos:
$ docker info
Containers: 21
 Running: 17
 Paused: 0
 Stopped: 4
Images: 57
Server Version: 1.10.3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 215
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: host bridge null
Kernel Version: 3.16.0-57-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.993 GiB
Name: xxxxxx
ID: xxxxxx
WARNING: No swap limit support
```
I hope this will help, thank you.

**Comment by SolidNerd** [comment]
@18601673727 Your Proxy Configuration is wrong if want to join a network that exists. In your case "nginx-proxy". It should be the following line: 
```
networks:
  nginx-proxy:
    external: true
  gitlab-tier:
    driver: bridge
```
I hope it helps.

**Comment by SolidNerd** [comment]
@18601673727 Here is my compose file: 
```
version: '2'
services:
  redis:
    restart: always
    image: sameersbn/redis:latest
    command:
    - --loglevel warning
    networks:
    - gitlab_backend
    volumes:
    - /srv/gitlab/redis:/var/lib/redis

  postgresql:
    restart: always
    image: sameersbn/postgresql:latest
    networks:
    - gitlab_backend
    volumes:
    - /srv/gitlab/postgresql:/var/lib/postgresql
    env_file:
    - gitlab_postgresql.env

  gitlab:
    restart: always
    image: sameersbn/gitlab:8.6.4
    networks:
    - gitlab_backend
    - nginx
    volumes:
    - /srv/gitlab/gitlab:/home/git/data
    ports:
    - "22:22"
    depends_on:
    - redis
    - postgresql
    env_file:
    - gitlab.env
networks:
  gitlab_backend:
  nginx:
    external: true
```

**Comment by 18601673727** [comment]
@SolidNerd Thanks for your kindly hinting but the result is exactly same as https://github.com/sameersbn/docker-gitlab/pull/658#issuecomment-207510689.

**Comment by 18601673727** [comment]
Good news, after upgraded Docker to version 1.11.1, SSH server will work even rebooting the containers with [this compose file](https://github.com/sameersbn/docker-gitlab/pull/658#issuecomment-208314359).

**Comment by 18601673727** [comment]
Hi @sameersbn, I just merged the current master into this branch and added a new file `docker-compose-v2.yml` to help keeping v2 users always be able to updating. 

Please take another look at this PR and merging would been perfect! = )

File: docker-compose-v2.yml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      - IMAP_USER=mailer@example.com
+      - IMAP_PASS=password
+      - IMAP_SSL=true
+      - IMAP_STARTTLS=false
+    volumes:
+      - /srv/docker/gitlab/gitlab:/home/git/data:Z
+    networks:
+      - gitlab-tier
+
+networks:
```
Comment (sameersbn) [inline docker-compose-v2.yml]: @18601673727 I think we should remove the `networks` block like @SolidNerd had suggested in the other PR. We can add it later if required.

Comment (SolidNerd) [inline docker-compose-v2.yml]: It`s only required when you run an external load-balancer like nginx or something else.

**Comment by 18601673727** [comment]
@sameersbn @SolidNerd Thanks for the notice, please have a look at the newest version 😄

**Comment by SolidNerd** [comment]
One last thing would be rebasing against master to fetch the latest updates.

**Comment by 18601673727** [comment]
@SolidNerd  What would you mean, this PR is already latest against master(62470c6c09d7cba4dd094cb09003a1ca4d4db904), isn't it?

**Comment by SolidNerd** [comment]
@18601673727 You are right it's already perfect.

**Comment by 18601673727** [comment]
@SolidNerd Thank you for the words and all the help here. :smile:

**Comment by sameersbn** [comment]
@18601673727 please rename the file to `docker-compose.yml`. we dont want to maintain another spec for docker-compose v1, only v2 going forward.

**Comment by 18601673727** [comment]
@sameersbn Got it, please check it out.

**Pull request merged**